spmd/compiler/fusion.py

```diff
@@ -0,0 +1,182 @@
+import logging
+from copy import deepcopy
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Iterable, List, Optional
+
+import torch
+import torch.fx as fx
+
+from .graph_utils import (
+    OP,
+    get_node_tensor_numel,
+    get_output_node,
+    graph_cleanup,
+    pretty_print_graph,
+)
+from .log_utils import rank0_debug
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+# enum for the supported fusion comm types
+class CommType(str, Enum):
+    allreduce = "allreduce_"
+    allgather = "allgather_"
+    broadcast = "broadcast_"
+    reducescatter = "reduce_scatter_"
+    scatter = "scatter_"
+
+
+@dataclass
+class FusionElement:
+    """This class tracks the nodes for a DTensor expanded communication collective in the graph"""
```

Comment:
Does this track all expanded collectives for the subgraphs as well? From the fe expansion list of ops, the pattern matching seemed to be for the dummy_add tensor alone.

Reply:
yes, at this point this only handles the dummy_add expansion, but is setup to be dynamic for any supported collective that uses the dummy_add insertion technique.